code/game/objects/items/reagent_containers/hypospray.dm

```diff
@@ -165,31 +121,51 @@
 	return TRUE
 
 /obj/item/reagent_containers/hypospray/afterattack_alternate(atom/A, mob/living/user)
+	if(can_draw_reagent(A, user, TRUE))
+		draw_reagent(A, user)
+		on_reagent_change()
+
+/obj/item/reagent_containers/hypospray/proc/can_draw_reagent(atom/A, mob/living/user, alternate_draw_mode)
 	if(!A.reagents)
-		return
+		return FALSE
 	if(!istype(user))
-		return
-	if(!in_range(A, user) || !user.Adjacent(A))
-		return
+		return FALSE
+	if(!(inject_mode == HYPOSPRAY_INJECT_MODE_DRAW || alternate_draw_mode)) //If we aren't drawing or using right click, return FALSE.
+		return FALSE
 	if(reagents.holder_full())
 		to_chat(user, span_warning("[src] is full."))
 		inject_mode = HYPOSPRAY_INJECT_MODE_INJECT
 		update_icon() //So we now display as Inject
-		return
-
-	if(istype(A, /obj)) //if not mob
+		return FALSE
+	if(iscarbon(A))
+		var/amount = min(reagents.maximum_volume - reagents.total_volume, amount_per_transfer_from_this)
+		var/mob/living/carbon/C = A
+		if(C.get_blood_id() && reagents.has_reagent(C.get_blood_id()))
+			to_chat(user, span_warning("There is already a blood sample in [src]."))
+			return
+		if(!C.blood_type)
+			to_chat(user, span_warning("You are unable to locate any blood."))
+			return
+		if(ishuman(C))
+			var/mob/living/carbon/human/H = C
+			if(H.species.species_flags & NO_BLOOD)
+				to_chat(user, span_warning("You are unable to locate any blood."))
+				return
+			else
+				C.take_blood(src,amount)
+		else
+			C.take_blood(src,amount)
+		reagents.handle_reactions()
+		user.visible_message("<span clas='warning'>[user] takes a blood sample from [A].</span>",
+							span_notice("You take a blood sample from [A]."), null, 4)
```

Comment:
This doesn't have anything to do with "can draw reagent". I would just make this a separate draw_blood proc or something.